### PR TITLE
xds: E2E Test for Audit Logging

### PR DESCRIPTION
**Pull request by gtcooke94**
This PR adds testing for audit logging logic to the existing RBAC tests. It very closely matches the tests for the authz path - https://github.com/grpc/grpc-go/pull/6304

I've currently only added a very basic test of the audit logger - the purpose here is to make sure that we can construct an audit logger through the xds path and it works.
I'm open to adding test cases for the various combinations of audit logging and authorization policies, but I think that's not really what we are testing here, and it's a pretty explosive combination and pollutes the test (if it's not necessary). All these cases should be covered in other tests like the unit tests and the authz path tests. @dfawley @easwars what do you think? I see compelling reasons for either choice.

RELEASE NOTES: N/A

### CODE REVIEW

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:635]: So we would just be changing around whether we are storing pointers/refs or just the actual object?

Comment (rockspore) [reply test/xds/xds_server_rbac_test.go:635]: No. It should still store pointers. But `cmp.Diff(nil, nil)` will be ok. So when neither `lastEvent` nor `test.eventContent` is populated the test should pass.

Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:635]: Discussed offline, we can't use `nil` because for the pointer to correctly propagate from the logger to the logger builder we do `*s.lastEvent = *lastEvent`, setting the underlying content being pointed to.

If `s.lastEvent` is `nil`, this causes a segfault, so we have to set it to an `audit.Event{}`

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:632]: This is not new but do you know why such a anonymous func is needed here?

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:701]: nit: remove extra empty line.

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:703]: Why is it called "unexpected message"? Maybe something like event content makes more sense?

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:981]: space before failed. same for the one below.

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:461]: It's blocked by another PR but I'd prefer the policy name, which would be the RBAC filter name, can be verified to be populated here.

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:632]: No clue, I'm just working into the existing test

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:703]: Changed to `Unexpected event`

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:981]: Good catch!

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:461]: I'll add a TODO

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:699]: Unless it's the name of the function, I don't think error messages should start with capital letters. Same below.

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:695]: I think it's worth considering adding some check to verify audit logging is not active till RBAC is toggled back.

**Review by rockspore — COMMENTED** [review]
Mostly LGTM. I'm also open to adding more test cases but I'll leave it to you and other reviewers.

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:695]: I think it ends up being somewhat implicit in the test - it makes 2 calls with RBAC on, then 2 calls with RBAC off, and the expected outcome for logging is `allowed: 2, denied: 0`

Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go:695]: Agree that this isn't the best as it requires holding some test state in your head as you look at it, but I hesitate to make large changes to this existing test

File: test/xds/xds_server_rbac_test.go
Comment (rockspore) [inline test/xds/xds_server_rbac_test.go:982]: Micro nit: I think should be lowercase `failed`. Same below.

**Review by rockspore — APPROVED** [review]

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Just one minor comment about the recoverable assertion check. LGTM otherwise.

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go]: Fixed

File: test/xds/xds_server_rbac_test.go
Comment (gtcooke94) [reply test/xds/xds_server_rbac_test.go]: Changed them to `Fatalf` - if the authz decisions don't match it doesn't make a lot of sense to try to compare the authz event outcomes because they are very likely nonsensical (because at a higher level the decision was wrong).

**Review by arvindbr8 — APPROVED** [review]

**Pull request merged**
